sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1386,15 +1386,17 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
 object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
   /**
    * Splits join condition expressions into three categories based on the attributes required
-   * to evaluate them.
+   * to evaluate them. Note that we explicitly exclude non-deterministic (i.e., stateful) condition
+   * expressions in canEvaluateInLeft or canEvaluateInRight to prevent pushing these predicates on
+   * either side of the join.
    *
    * @return (canEvaluateInLeft, canEvaluateInRight, haveToEvaluateInBoth)
    */
   private def split(condition: Seq[Expression], left: LogicalPlan, right: LogicalPlan) = {
     val (leftEvaluateCondition, rest) =
-        condition.partition(_.references subsetOf left.outputSet)
+        condition.partition(expr => expr.references.subsetOf(left.outputSet) && expr.deterministic)
```

Comment:
Maybe we need to use `span` here. See another PR: https://github.com/apache/spark/pull/14012